internal/cephfs/controllerserver.go

```diff
@@ -36,25 +41,111 @@ type ControllerServer struct {
-	cr, err := util.NewAdminCredentials(secret)
-	if err != nil {
-		return status.Error(codes.InvalidArgument, err.Error())
+func (cs *ControllerServer) createBackingVolume(
+	ctx context.Context,
+	volOptions *volumeOptions,
+	vID *volumeIdentifier,
+	parentVolOpt *volumeOptions,
+	pvID *volumeIdentifier,
+	sID *snapshotIdentifier,
+	cr *util.Credentials) error {
+	var err error
+	if sID != nil {
+		snapID := volumeID(sID.FsSnapshotName)
+		err = cloneSnapshot(ctx, parentVolOpt, cr, volumeID(pvID.FsSubvolName), snapID, volumeID(vID.FsSubvolName), volOptions)
+		if err != nil {
+			return err
+		}
+		defer func() {
+			if err != nil {
+				if dErr := purgeVolume(ctx, volumeID(vID.FsSubvolName), cr, volOptions); dErr != nil {
+					klog.Errorf(util.Log(ctx, "failed to delete volume %s: %v"), vID.FsSubvolName, dErr)
+				}
+			}
+		}()
+		// This is a work around to fix sizing issue for cloned images
+		err = resizeVolume(ctx, volOptions, cr, volumeID(vID.FsSubvolName), volOptions.Size)
+		if err != nil {
+			klog.Errorf(util.Log(ctx, "failed to expand volume %s: %v"), vID.FsSubvolName, err)
+			return status.Error(codes.Internal, err.Error())
+		}
+		var clone = CloneStatus{}
+		clone, err = getcloneInfo(ctx, volOptions, cr, volumeID(vID.FsSubvolName))
+		if err != nil {
+			return status.Error(codes.Internal, err.Error())
+		}
+		if clone.Status.State != cephFSCloneCompleted {
+			return ErrCloneInProgress{err: fmt.Errorf("clone is in progress for %v", vID.FsSubvolName)}
+		}
+		return nil
+	}
+
+	if parentVolOpt != nil {
+		err = createCloneFromSubvolume(ctx, volumeID(pvID.FsSubvolName), volumeID(vID.FsSubvolName), volOptions, parentVolOpt, cr)
+		if err != nil {
+			return status.Error(codes.Internal, err.Error())
+		}
+		return nil
 	}
-	defer cr.DeleteCredentials()
 
 	if err = createVolume(ctx, volOptions, cr, volumeID(vID.FsSubvolName), volOptions.Size); err != nil {
 		klog.Errorf(util.Log(ctx, "failed to create volume %s: %v"), volOptions.RequestName, err)
 		return status.Error(codes.Internal, err.Error())
 	}
-
 	return nil
 }
 
+func checkContentSource(ctx context.Context, req *csi.CreateVolumeRequest, cr *util.Credentials) (*volumeOptions, *volumeIdentifier, *snapshotIdentifier, error) {
+	if req.VolumeContentSource == nil {
+		return nil, nil, nil, nil
+	}
+	volumeSource := req.VolumeContentSource
+	switch volumeSource.Type.(type) {
+	case *csi.VolumeContentSource_Snapshot:
+		snapshotID := req.VolumeContentSource.GetSnapshot().GetSnapshotId()
+		var snof util.ErrSnapNotFound
+		snapOpt, _, sid, err := newSnapshotOptionsFromID(ctx, snapshotID, cr)
+		if err != nil {
+			if errors.As(err, &snof) {
+				return nil, nil, nil, status.Error(codes.NotFound, err.Error())
+			}
+			return nil, nil, nil, status.Error(codes.Internal, err.Error())
+		}
+		_, _, err = checkSnapExists(ctx, snapOpt, sid.FsSubvolName, sid.FsSnapshotName, cr)
```

Comment:
`newSnapshotOptionsFromID` already checks and processes snapshot info, why the checkSnapExists again?

Further, garbage collecting snapshot during a clone does not seem appropriate, if the snapshot was half baked, then the CO cannot have a snapshot ID to pass as source.